godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,480 @@
+        // usage is unclear
+        set_fallback_func: None,
+    };
+
+    let data = ScriptInstanceData {
+        inner: RefCell::new(rs_instance),
+    };
+
+    unsafe {
+        let data_ptr = Box::into_raw(Box::new(data));
+        let instance_ptr = Box::into_raw(Box::new(gd_instance));
+
+        get_interface().script_instance_create.unwrap()(
+            instance_ptr,
+            data_ptr as GDExtensionScriptInstanceDataPtr,
+        )
+    }
+}
+
+mod script_instance_info {
+    use std::{
+        ffi::c_void,
+        mem,
+        ptr::drop_in_place,
+        slice::{self},
+    };
+
+    use godot_ffi::{
+        force_mut_ptr, GDExtensionBool, GDExtensionCallError, GDExtensionConstStringNamePtr,
+        GDExtensionConstVariantPtr, GDExtensionInt, GDExtensionMethodInfo, GDExtensionObjectPtr,
+        GDExtensionPropertyInfo, GDExtensionScriptInstanceDataPtr, GDExtensionScriptLanguagePtr,
+        GDExtensionStringPtr, GDExtensionTypePtr, GDExtensionVariantPtr, GDExtensionVariantType,
+        GodotFfi, PtrcallType, GDEXTENSION_CALL_OK,
+    };
+
+    use crate::builtin::{meta::GodotType, StringName, Variant};
+
+    use super::{ScriptInstance, ScriptInstanceData};
+
+    pub(super) unsafe extern "C" fn set_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        p_name: GDExtensionConstStringNamePtr,
+        p_value: GDExtensionConstVariantPtr,
+    ) -> GDExtensionBool {
+        let instance = p_instance as *mut ScriptInstanceData<T>;
+        let name = StringName::from_arg_ptr(p_name as GDExtensionTypePtr, PtrcallType::Standard);
+        let value = &*Variant::ptr_from_sys(force_mut_ptr(p_value));
+
+        let result = (*instance).inner.borrow_mut().set(name.clone(), value) as u8;
+
+        result
+    }
+
+    pub(super) unsafe extern "C" fn get_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        p_name: GDExtensionConstStringNamePtr,
+        r_ret: GDExtensionVariantPtr,
+    ) -> GDExtensionBool {
+        let instance = p_instance as *const ScriptInstanceData<T>;
+        let name = StringName::from_arg_ptr(p_name as GDExtensionTypePtr, PtrcallType::Standard);
+
+        let result = (*instance)
+            .inner
+            .borrow()
+            .get(name.clone())
+            .map(|variant| {
+                variant.move_return_ptr(
+                    r_ret as GDExtensionTypePtr,
+                    godot_ffi::PtrcallType::Standard,
+                )
+            })
+            .is_some();
+
+        result as u8
+    }
+
+    pub(super) unsafe extern "C" fn get_property_list_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        r_count: *mut u32,
+    ) -> *const GDExtensionPropertyInfo {
+        let instance = p_instance as *const ScriptInstanceData<T>;
+
+        let property_list: Vec<_> = (*instance)
+            .inner
+            .borrow()
```

Comment:
These lines are a recurring pattern in the C functions -> could be extracted into a function that returns the guard. `borrow()` will still be needed on the call site, for RAII.